File: 0118-pascals-triangle/0118-pascals-triangle.java
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> row0 = new ArrayList<>(), row1 = new ArrayList<>();
        row0.add(1);
        res.add(row0);

        for (int i = 1; i < numRows; i++) {
            List<Integer> currRow = new ArrayList<>();
            currRow.add(1);
            for (int j = 1; j < i; j++) {
                currRow.add(res.get(i - 1).get(j) + res.get(i - 1).get(j - 1));
            }
            currRow.add(1);
            res.add(currRow);
        }
        
        return res;
    }
}

File: 0118-pascals-triangle/0118-pascals-triangle.cpp
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> res;
        res.push_back({1});
        
        for (int i = 2; i <= numRows; i++) {
            vector<int> curr_row(i);
            curr_row[0] = 1;
            curr_row[i - 1] = 1;
            vector<int>& prev_row = res[i - 2];
            for (int j = 1; j < i - 1; j++) {
                curr_row[j] = prev_row[j - 1] + prev_row[j];
            }
            res.push_back(curr_row);
        }
        
        return res;
    }
};